### PR TITLE
Add an n_event_bunch to sampler classes to alleviate memory issues when sampling bright sources

### PR DESCRIPTION
**Pull request by registerrier**
**Description**

This pull request proposes a solution to #5112 . 
The issue there is that IRF sampling requires large memory consumption which is proportional to the number of events sampled. This can become problematic for some bright sources.

The proposed solution is to introduce a slicing mechanism into `IRFMap.sample_coord()` to perform sampling by bunches of fixed and controled sizes. 
By default, the behavior is unchanged (sample all events at once), but a parameter can be set on `MapDatasetEventSampler` to limit the chunk size.

**Dear reviewer**

### CODE REVIEW

File: gammapy/datasets/simulate.py [hunk 1]
```diff
@@ -568,11 +577,13 @@ def run(self, dataset, observation=None):
         events_bkg = self.sample_background(dataset)
         events = EventList.from_stack([events_bkg, events_src])
 
-        events = self.event_det_coords(observation, events)
         events.table["EVENT_ID"] = np.arange(len(events.table))
-        events.table.meta.update(
-            self.event_list_meta(dataset, observation, self.keep_mc_id)
-        )
+        if observation is not None:
+            print("det coords")
```
Comment (adonath) [inline gammapy/datasets/simulate.py:582]:
```suggestion
```

**Review by adonath — DISMISSED** [review]
Thank @registerrier! This is indeed a very good solution, we can just rely on Python's memory handling. When you initially proposed the fix, I immediately thought about grouping events into spatial bins. But this is more for overall performance than for memory handling. No further comments from my side.

File: gammapy/datasets/simulate.py [hunk 2]
```diff
@@ -37,6 +37,9 @@ class MapDatasetEventSampler:
     keep_mc_id : bool, optional
         Flag to tag sampled events from a given model with a Montecarlo identifier.
         Default is True. If set to False, no identifier will be assigned.
+    n_event_bunch : int, optional
+        Size of events bunches to sample. If None, sample all events in memory.
+        Default is None.
```
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/datasets/simulate.py:42]: I think the right number here is the smallest number, such that the Python loop does not become the bottleneck. This offers the smaller memory footprint, while not affecting performance. However there I snot single answer to this. 

@registerrier I you don't mind you could maybe do a quick plot for runtime vs chunk size, using sizes of 8, 16, 32, 64, 128,  etc. and see how this scales.

[the remark this answers is not in the document]
Comment (registerrier) [reply gammapy/datasets/simulate.py:42]: I have put a default at `None` to have identical backward behaviour. But it probably does not make sense, as this is supposed to solve an issue :) .
I think 1e4 is probably a good typical value with current IRFs. See below.

**Comment by adonath** [comment]
Thanks @registerrier, 1e4` looks like a reasonable default!

**Comment by fabiopintore** [comment]
Thanks @registerrier . No further comments from my side.

**Pull request merged**